packages/suite/src/views/onboarding/UnexpectedState/index.tsx

```diff
@@ -60,33 +65,35 @@ const UnexpectedState = ({ children }: UnexpectedStateProps) => {
         }
     }, [activeStep, prerequisite, isNotSameDevice]);
 
-    const getPinComponent = () => {
-        // After the PIN is set it may happen that it takes too long for an user to finish the onboarding process.
-        // Then the device will get auto locked and requests to show a PIN matrix next before changing its setting.
-        // (which could happen on Final step where we set device name and homescreen)
-        if (!device?.features) return null;
-        if (activeStepId === 'set-pin') return null; // Step for setting up a PIN handles all by itself
-        if (showPinMatrix) {
-            return <PinMatrix device={device} />;
-        }
+    const handlePinSubmit = () => {
+        dispatch(onPinSubmit(pin));
+        setPin('');
     };
 
     if (!activeStep) {
         return null;
     }
 
-    const pinComponent = getPinComponent();
-    if (pinComponent) {
+    // After the PIN is set it may happen that it takes too long for an user to finish the onboarding process.
+    // Then the device will get auto locked and requests to show a PIN matrix next before changing its setting.
+    // (which could happen on Final step where we set device name and homescreen)
+    if ((!device?.features || activeStepId === 'set-pin') && showPinMatrix) {
```

Comment:
idk how to reproduce the so called Unexpected States, but moving the condition here from getPinComponent is much nicer :+1: 
:point_right:  But I believe you forgot to negate the `(!device?.features || activeStepId === 'set-pin')`
(if either is truth, then it should be null -> don't render this).

Though I'd invert that negation to simplify it:
```suggestion
    if (device?.features && activeStepId !== 'set-pin' && showPinMatrix) {
```